ethcore/sync/src/chain/handler.rs

```diff
@@ -87,9 +87,26 @@ impl SyncHandler {
 				Ok(())
 			}
 		};
-		result.unwrap_or_else(|e| {
-			debug!(target:"sync", "{} -> Malformed packet {} : {}", peer, packet_id, e);
-		})
+
+		match result {
+			Err(DownloaderImportError::Invalid) => {
+				debug!(target:"sync", "{} -> Malformed packet {}", peer, packet_id);
+				io.disable_peer(peer);
+				sync.deactivate_peer(io, peer);
+				sync.continue_sync(io);
+			},
+			Err(DownloaderImportError::Useless) => {
+				sync.deactivate_peer(io, peer);
+				// give tasks to other peers
+				sync.continue_sync(io);
+			},
+			Ok(()) => {
+				// give a task to the same peer first
+				sync.sync_peer(io, peer, false);
```

Comment:
There are some methods where `sync_peer` is still called inside (both with force `true` and `false`). I guess this will be a reason for some extra `Skipping busy peer` messages in the log (since now it is called twice - inside methods and in this match). Not an issue, though :)

Reply:
Thank's for pointing out. I'll remove all remaining `sync.sync_peer(io, peer, false)` :)